### ISSUE TITLE
softmaxCrossEntropy can't be used directly as a loss function

### ISSUE THREAD

**Issue by sgugger** [tensorflow/swift-apis#422]
Because it has a third argument (reduction) with a default value, ` softmaxCrossEntropy` can no longer be used as a loss function in the `Learner` API of the harebrain notebooks. A workaround is to define a new function that has only two arguments and pass it:
```
@differentiable(wrt: logits)
func crossEntropy(logits: Tensor<Float>, labels: Tensor<Int32>) {
    return softmaxCrossEntropy(logits: logits, labels: labels
}
```
but it's a bit of a waste of code...

**Comment by Shashi456** [comment]
@sgugger so this decision was taken in the #356 PR, in an attempt to standardize loss functions. The reduction was added to maintain source compatibility.

**Comment by sgugger** [comment]
I think it's a good idea to have the reduction parameter in the loss function in theory. However, it's very annoying that swift doesn't consider that a function with three arguments, the third one having a default value, can be interpreted as a function of two arguments.

**Comment by rxwei** [comment]
@Shashi456 Thanks for chiming in to provide the context. The core issue here is not about whether we should have a reduction argument, but Swift's lack of support for function coercion with default parameters. Since reduction is a default parameter in all loss functions, ideally we should be able to coerce them to a function type that excludes the defaulted parameter.
```swift
func foo(x: T, y: T = z) -> T { ... }
foo as (T) -> T // Nope, Swift does not support it.
```

Right now there is a workaround: anywhere you passed `softmaxCrossEntropy`, you pass `{ softmaxCrossEntropy(logins: $0, labels: $1) }` instead. This is less heavyweight than declaring a new function. cc @sgugger 

To fix the breakage in the short term, we can consider making the `reduction` parameter not be a default parameter, but declare an overload that does not take a `reduction` but instead passes `mean` in, and do this for all loss functions. @eaplatanios what do you think?

To keep discussions going, we should start a discussion on the [Swift Evolution forum](https://forums.swift.org/c/evolution) about coercing functions with default parameters.

**Comment by Shashi456** [comment]
@rxwei, would there be any way to do that generically for all the losses ?

**Comment by rxwei** [comment]
> @rxwei, would there be any way to do that generically for all the losses ?
> Like we made a function layer for differentiable functions.

There is not, unfortunately.

**Comment by eaplatanios** [comment]
We could take the approach Richard suggests and add overloads. I am personally leaning a bit more towards not adding the overloads in the `TensorFlow` module and letting users define them as needed. This is for a couple reasons:

1. The workaround is quite simple and would not add much overhead on the library user.
2. The high-level API is currently in early design form and will be changing more in the future. In this case I believe clearly documenting workarounds and changes is a better approach than adding overloads for backwards compatibility so that we don't obfuscate the `TensorFlow` module. Backwards source compatibility should become a bigger concern later down the line I believe.

However, my opinion here is not strong and I'm fine going either way. :)

@rxwei starting a discussion on coercing functions with default parameters sounds great. I guess this is also tightly connected to currying for functions with non-default parameters and it may be useful to bring both of these up in that discussion. :)

**Comment by Shashi456** [comment]
@eaplatanios +1, on your thoughts. 

But if it is important for us to overload them in the library, I would then suggest moving the reducer functions to Math.swift while having the overloaded functions in loss.swift.

**Comment by dynamicwebpaige** [comment]
Pinging on this issue, as well -- do we have an ETA to resolve? cc: @saeta

**Comment by Shashi456** [comment]
If it's overloading the current losses, or just `softmaxcrossentropy` specifically, I could get a PR up today.

**Comment by saeta** [comment]
I think we should add the overloads; it's now just a matter of how we want to organize the code. 

To make forward progress, how about this strawperson proposal:
 1. Leave existing losses where they are.
 2. Add additional overloads without the 3rd param
 3. Optionally in a subsequent PR / discussion, determine if we should move them to Math.swift. (My initial reaction is that putting them in math makes sense, but I can definitely see both sides (esp in light of step 5). I'd be very interested in hearing what other folks think.)
 4. Propose adding function coercion to Swift & do the implementation.
 5. Remove the extra overloads.

 My rationale for adding the overloads is that we have caused a regression in the fastai tutorials. (In general, I'm quite sympathetic to keeping the standard library small (@eaplatanios articulates this well), but I think the end-to-end usage demonstrated by the fastai tutorials indicates that having the overloads in the API repo is ergonomically preferable.)

What do folks think? @Shashi456 would you be interested in putting together a PR (subject to consensus on this proposal)?

**Comment by Shashi456** [comment]
I'll get the PR immediately up with the overloaded losses once we have consensus @saeta , Also what do you think about naming these functions? 
Step 3 sounds like a sound thing to do in given we are going to keep both versions around. This is also what @eaplatanios originally proposed in #356. I'm not entirely sure how quick swift is as far as adding new functionalities to the language are considered, but whenever they are in place, It'd be a good move forward for this library as well.

**Comment by eaplatanios** [comment]
@saeta @Shashi456 How about we add the overloads in `BackwardsCompatibility.swift` which was created for the purpose of keeping track of extra APIs that are mostly temporary and exist for the sake of maintaining backwards compatibility?

**Comment by saeta** [comment]
:+1:

**Comment by Shashi456** [comment]
@eaplatanios @saeta, I'll get the PR up in a few hours.

**Comment by rahulbhalley** [comment]
@rxwei function coercion looks like a great 👍 idea 💡 to me. Please do make a proposal in Swift evolution forum.

**Comment by Shashi456** [comment]
@sgugger @dynamicwebpaige solved this in #442.
